pypicloud/access/_ldap.py

```diff
@@ -0,0 +1,391 @@
+
+        if not hasattr(LDAP, "_all_users"):
+            results = LDAP._server.search_s(
+                LDAP._base_dn,
+                ldap.SCOPE_SUBTREE,
+                LDAP._all_user_search,
+            )
+            LDAP._all_users = {}
+            for result in results:
+                if LDAP._id_field in result[1]:
+                    LDAP._all_users[result[1][LDAP._id_field][0]] = result[0]
+
+        return list(LDAP._all_users.values())
+
+    @staticmethod
+    def all_usernames():
+        """Returns a list of all user names."""
+
+        return list(LDAP._all_users.keys())
+
+    @staticmethod
+    @reconnect
+    def _add_admins_from_dn(admin_dn):
+        """Given a DN fragement, add users to _admins and _admin_usernames."""
+
+        res = LDAP._server.search_s(admin_dn, ldap.SCOPE_SUBTREE)
+        try:
+            admins = res[0][1][LDAP._admin_field]
+        except (IndexError, KeyError) as err:
+            logging.warn("Error retrieving admins from %s: %r", admin_dn, err)
+            return
+
+        for admin in admins:
+            try:
+                LDAP._admin_usernames.append(
+                    LDAP._server.search_s(
+                        admin,
+                        ldap.SCOPE_SUBTREE,
+                    )[0][1][LDAP._id_field][0]
+                )
+            except (IndexError, KeyError) as error:
+                logging.warn("Error looking up admin %s: %r", admin, error)
+            else:
+                LDAP._admins.append(admin)
+
+    @staticmethod
+    def admins():
+        """Returns a list of all the admin DNs."""
+
+        if not hasattr(LDAP, "_admins"):
+            LDAP._admins = []
+            LDAP._admin_usernames = []
+            for admin_dn in LDAP._admin_dns:
+                LDAP._add_admins_from_dn(admin_dn)
+
+        return LDAP._admins
+
+    @staticmethod
+    def admin_usernames():
+        """Returns a list of the admin usernames."""
+
+        if not hasattr(LDAP, "_admin_usernames"):
+            LDAP.admins()
+        return LDAP._admin_usernames
+
+    @staticmethod
+    def user_dn(username):
+        """Returns the dn for the username."""
+
+        return LDAP._all_users.get(username)
+
+    @staticmethod
+    @reconnect
+    def bind_user(user_dn, password):
+        """Attempts to bind as the user, then rebinds as service user again."""
+
+        try:
+            LDAP._server.simple_bind_s(user_dn, password)
+        except ldap.INVALID_CREDENTIALS:
+            return False
+        else:
+            return True
+        finally:
+            LDAP._connect()
+
+
+class LDAPAccessBackend(IAccessBackend):
+    """
+    This backend allows you to store all user and package permissions in a SQL
+    database and authenticate against a LDAP server
+    """
+
+    DEFAULT_PERMISSIONS = {
+        "admin": ("read", "write"),
+        "authenticated": ("read", "write"),
+        "everyone": ("read",),
+    }
```

Comment:
In general I would prefer that the user configure this with the `pypi.default_read` and `pypi.default_write` options. Unless I misunderstand, this makes it completely inflexible, so if someone wants logged-out users to be unable to download packages, there is no way to do that. Is there a good reason to do the configuration this way?